src/uwtools/template.py

```diff
@@ -0,0 +1,172 @@
+        build_index to create a dictionary that will have all the options from deeper levels (list, dicts).
+        You can then pass index.get as an argument to any method you use.
+        If you use substitute_with_dependencies, this is done automatically.
+    """
+
+    substitutions = {
+        TemplateConstants.DOLLAR_CURLY_BRACE: TemplateConstants.SubPair(re.compile('\${.*?}+'), slice(2, -1)),
+        TemplateConstants.DOLLAR_PARENTHESES: TemplateConstants.SubPair(re.compile('\$\(.*?\)+'), slice(2, -1)),
+        TemplateConstants.DOUBLE_CURLY_BRACES: TemplateConstants.SubPair(re.compile('{{.*?}}+'), slice(2, -2)),
+        TemplateConstants.AT_SQUARE_BRACES: TemplateConstants.SubPair(re.compile('@\[.*?\]+'), slice(2, -1)),
+        TemplateConstants.AT_ANGLE_BRACKETS: TemplateConstants.SubPair(re.compile('@\<.*?\>+'), slice(2, -1))
+    }
+
+    def is_single_type_or_string(s):
+        if isinstance(s, str):
+            return True
+        try:
+            len(s)
+        except TypeError:
+            return True
+        else:
+            return False    
+
+    @classmethod
+    def find_variables(cls, variable_to_substitute: str, var_type: str):
+        pair = cls.substitutions[var_type]
+        return [x[pair.slice] for x in re.findall(pair.regex, variable_to_substitute)]
+
+    @classmethod
+    def substitute_string(cls, variable_to_substitute, var_type: str, get_value):
+        """
+            Substitutes variables under the form var_type (e.g. DOLLAR_CURLY_BRACE), looks for a value returned
+            by function get_value and if found, substitutes the variable. Convert floats and int to string
+            before substitution. If the value in the dictionary is a complex type, just assign it instead
+            of substituting.
+                get_value is a function that returns the value to substitute:
+                signature: get_value(variable_name).
+                If substituting from a dictionary my_dict,  pass my_dict.get
+        """
+        pair = cls.substitutions[var_type]
+        if isinstance(variable_to_substitute, str):
+            variable_names = re.findall(pair.regex, variable_to_substitute)
+            for variable in variable_names:
+                var = variable[pair.slice]
+                v = get_value(var)
+                if v is not None:
+                    if not cls.is_single_type_or_string(v):
+                        if len(variable_names) == 1:
+                            # v could be a list or a dictionary (complex structure and not a string).
+                            # If there is one variable that is the whole
+                            # string, we can safely replace, otherwise do nothing.
+                            if variable_to_substitute.replace(variable_names[0][pair.slice], '') == var_type:
+                                variable_to_substitute = v
+                    else:
+                        if isinstance(v, float) or isinstance(v, int):
```

Comment:
```if isinstance(v, (float, int)):```
https://stackoverflow.com/questions/33311258/python-check-if-isinstance-any-type-in-list